### ISSUE TITLE
Base image development encounters load metadata errors (darwin-arm64)

### ISSUE THREAD

**Issue by colelawrence** [tilt-dev/tilt#4932]
### Expected Behavior

Should be able to build base images from all images.

Using k3d, macOS M1 aarch64, running `tilt up` in the following repo:
https://github.com/colelawrence/nodejs-express-k8s/commit/33b9e6418b6b03c1572283aee98affede4e7d924

### Current Behavior

```
STEP 1/5 — Building Dockerfile: [tilt.dev/nodejs-express-base]
Building Dockerfile:
  FROM node:14-alpine
  
  WORKDIR '/var/www/app'
  ADD package.json package.json
  RUN npm install
  ENTRYPOINT node server.js


     Tarring context…
     Building image
     copy /context / [done: 45ms]
     [1/4] FROM docker.io/library/node:14-alpine@sha256:8c94a0291133e16b92be5c667e0bc35930940dfa7be544fb142e25f8e4510a45
     [2/4] WORKDIR /var/www/app [cached]
     [3/4] ADD package.json package.json [cached]
     [4/4] RUN npm install [cached]
     exporting to image

STEP 2/5 — Pushing localhost:52780/tilt.dev_nodejs-express-base:tilt-858fb587e21f53c7
     Skipping push: base image does not need deploy

STEP 3/5 — Building Dockerfile: [tilt.dev/nodejs-express-app]
Building Dockerfile:
  FROM localhost:52780/tilt.dev_nodejs-express-base:tilt-858fb587e21f53c7
  
  WORKDIR '/var/www/app'
  
  ADD . .


     Tarring context…
     Building image
     copy /context / [done: 43ms]
     
     ERROR IN: [internal] load metadata for localhost:52780/tilt.dev_nodejs-express-base:tilt-858fb587e21f53c7

Build Failed: ImageBuild: failed to create LLB definition: localhost:52780/tilt.dev_nodejs-express-base:tilt-858fb587e21f53c7: not found
```

### Steps to Reproduce

https://github.com/colelawrence/nodejs-express-k8s/commit/33b9e6418b6b03c1572283aee98affede4e7d924

0. Be using `k3d cluster create --config $configfile` with a create registry on latest Docker for Desktop for macOS M1.
1. `git clone https://github.com/colelawrence/nodejs-express-k8s`
2. `tilt up`

#### `tilt doctor` Output

```
$ tilt doctor
Tilt: v0.22.7, built 2021-09-03
System: darwin-arm64
---
Docker
- Host: [default]
- Server Version: 20.10.8
- API Version: 1.41
- Builder: 2
---
Kubernetes
- Env: k3d
- Context: k3d-tilt
- Cluster Name: k3d-tilt
- Namespace: default
- Container Runtime: containerd
- Version: v1.21.3+k3s1
- Cluster Local Registry: {Host:localhost:52780 hostFromCluster:k3d-tilt-registry:5000 SingleName:}
---
Thanks for seeing the Tilt Doctor!
Please send the info above when filing bug reports. 💗

The info below helps us understand how you're using Tilt so we can improve,
but is not required to ask for help.
---
Analytics Settings
--> (These results reflect your personal opt in/out status and may be overridden by an `analytics_settings` call in your Tiltfile)
- User Mode: opt-out
- Machine: eebd79741ae00518360169d555d4ce7e
- Repo: gCGWHHiUiupQMoOekqknDQ==
```

#### About Your Use Case

I cannot successfully use base images with `tilt up`.

(In my _real_ code, I actually was experiencing this `ERROR IN: [internal] load metadata for localhost:52780/tilt.dev_nodejs-express-base:tilt-858fb587e21f53c7` when I was based from `rust:buster`, so I'm suspicious that it's some kind of syntax property or build kit issue... maybe related to aarch64 (since that's usually the issue I end up hitting).

I've been debugging this for about 3 hours trying to figure out what the difference was between the example and my code. I really don't know how to pry into the `ERROR IN: [internal] load metadata for` error, but I am happy to help figure it out.

**Comment by colelawrence** [comment]
I have a suspicion that this is related to when the docker image is published to hub.docker.io. Perhaps there is some kind of issue with the k3d registry?

* `node:14.17.5-alpine3.13` from 23 days ago works

while 
* `node:14.17.5-alpine3.14` from 8 days ago DOES NOT work
* `node:14.17.6-alpine3.14` from 4 days ago DOES NOT work

**Comment by colelawrence** [comment]
I've been trying an assortment of additional little things, and I'm not so sure anymore what the issue is... I wish I knew how to troubleshoot this more deeply than just getting `ERROR IN: [internal] load metadata for ...`.

**Comment by colelawrence** [comment]
It appears that multi-stage builds are a whole other complexity with this error at it's center

**Comment by colelawrence** [comment]
I've opened another branch on my repro repo. This demonstrates a little more complex of a set-up with multiple-stage builds which fail with similar issues.

I'm not sure if the two issues are a part of the same issue, but both prevent me from using tilt with both multi-stage and base images.

For the record, independent multi-stage images work fine even pointing against latest tags.

https://github.com/tilt-dev/nodejs-express-k8s/commit/70d2bc4c2467e7a7437fddc44fef1ed6e1878e3d

**Comment by nicks** [comment]
I tried your example on Linux/Kind and Linux/k3d, and it built OK for me. We can poke around on it more after the weekend.

Given some of the symptoms you're describing, in particular:
-  node:14.17.5-alpine3.13 works but node:14.17.5-alpine3.14 does not work
- weird errors about the image not being found
- the fact that you're on a MacOS M1

my educated guess is you've run into some sort of multi-arch image bug (i.e., you're dealing with an image for multiple chipsets, and some part of the system is handling it incorrectly)

**Comment by colelawrence** [comment]
@nicks, thanks so much! I'll be happy to help diagnose when the weekdays come. For now, I might just have to switch operating systems, or I might see if I can force building a more universal byte code with qemu.

**Comment by colelawrence** [comment]
I've begun testing with `kind` with `ctlptl`.

I'm largely experiencing the exact same issues. It appears that this issue happens even without Tilt or k8s.

I've created a few minimal test cases at https://github.com/colelawrence/nodejs-express-k8s/blob/7923b335a4dcc3689fcd800e47122fba56c04d82

In this repo, I have a set of `x-*.sh` files for tests. Each one starts by `docker system prune -a -f` and then performs a few `docker build` commands.
At [the head of every log output](
https://github.com/colelawrence/nodejs-express-k8s/blob/7923b335a4dcc3689fcd800e47122fba56c04d82/x-fail-docker-build.sh-arm64.log#L1-L18), I print the contents of the script followed by the stderr + stdout output of the script.

The gist so far, is that the commands work when I `export DOCKER_DEFAULT_PLATFORM=linux/arm64`.

This may have an effect on several things, but one thing I found very interesting was this `git diff` between fail and pass for the `docker image inspect tilt.dev/demo-base | jq` command:

![image](https://user-images.githubusercontent.com/2925395/132133528-8fed6a5d-d406-404b-a5b3-6b3bb26bea9e.png)

The passing version does NOT have the "Variant": "v8". I'm not sure if this is related, but it seems interesting.

**Comment by colelawrence** [comment]
Possibly related https://github.com/docker/for-mac/issues/5625

**Comment by colelawrence** [comment]
This looks _very_ related https://github.com/docker/for-mac/issues/5873#issuecomment-895724525

**Comment by colelawrence** [comment]
One thing that I think could immediately unblock me is if I could either set docker settings via env vars for `docker_build` (so I can set `--platform linux/arm64` or I can specify a `DOCKER_DEFAULT_PLATFORM=linux/arm64` that would be reflected in my Tiltfile `docker_build` settings. Otherwise, I am actually able to do `custom_build` where I manually set `--platform`, but I end up missing out on the live_update tools, I think.

**Comment by nicks** [comment]
Does setting DOCKER_BUILDKIT=0 fix the problem? Tilt should respect that.

**Comment by colelawrence** [comment]
Setting `DOCKER_BUILDKIT=0` fixes the minimal repro [(see logs)](https://github.com/colelawrence/nodejs-express-k8s/blob/2ac02ef324e3e880f6232860c2485ca57f3cdaa8/x-pass-docker-build-with-no-builderkit.sh-arm64.log), but I rely on it for the `--mount` features in one of my images.

I can look into re-writing images that were requiring `--mount`.

**Comment by colelawrence** [comment]
I do notice that I have to prune between tests, if I don't, then setting `DOCKER_BUILDKIT=0` still fails (I suspect its still using images built using BUILDKIT which still don't match up)

**Comment by milas** [comment]
Thanks for all the debugging @colelawrence!

I just opened #4934 which will allow you to either explicitly pass `platform` as an arg to `docker_build` or respect `DOCKER_DEFAULT_PLATFORM` if not explicitly set (i.e. matching the Docker CLI behavior).

You could handle this automatically by putting something like the following at the beginning of your `Tiltfile` after upgrading to the next release once it's out:
```python
if 'DOCKER_DEFAULT_PLATFORM' not in os.environ:
    arch=str(local('uname -m', command_bat='echo %PROCESSOR_ARCHITECTURE%', quiet=True, echo_off=True)).rstrip().lower()
    if arch == 'arm64':
        print('Setting default architecture for Docker builds to "linux/amd64"')
        os.putenv('DOCKER_DEFAULT_PLATFORM', 'linux/arm64')
```

**Comment by milas** [comment]
[v0.22.8+](https://github.com/tilt-dev/tilt/releases/latest) is out which supports `platform` for `docker_build` - if you run into any problems with it, feel free to comment here or open a new issue.

See https://docs.tilt.dev/api.html#api.docker_build for usage
